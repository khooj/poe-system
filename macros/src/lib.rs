use std::{
    env, fs,
    io::{BufRead, BufReader},
};

use proc_macro::TokenStream;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input,
    token::Comma,
    Ident, LitStr,
};

struct StaticMacro {
    array_name: Ident,
    filename: String,
}

impl Parse for StaticMacro {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let name: Ident = input.parse()?;
        let _: Comma = input.parse()?;
        let filename: LitStr = input.parse()?;

        Ok(StaticMacro {
            array_name: name,
            filename: filename.value(),
        })
    }
}

#[proc_macro]
pub fn static_array_from_file(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as StaticMacro);
    let name = input.array_name;
    let filename = input.filename;

    let mut c = env::current_dir().unwrap();
    c.push(&filename);
    let f = fs::OpenOptions::new().read(true).open(&c);
    if f.is_err() {
        let e = format!("cant open file: {}", c.to_string_lossy());
        return quote!(::std::compile_error!(#e);).into();
    }

    let mut buf = BufReader::new(f.unwrap());
    let mut sv = vec![];
    let mut s = String::new();
    while let Ok(sz) = buf.read_line(&mut s) {
        if sz == 0 {
            break;
        }
        sv.push(s.clone());
        s.clear();
    }

    let tokens = quote! {
        static #name: &[&'static str] = &[ #(#sv),* ];
    };

    tokens.into()
}
