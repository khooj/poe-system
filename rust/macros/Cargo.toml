[package]
name = "macros"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
syn = "1"
quote = "1"
proc-macro2 = "1"

[dev-dependencies]
trybuild = "1.0"
