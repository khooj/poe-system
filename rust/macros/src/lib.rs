use std::{
    env, fs,
    io::{BufRead, BufReader},
};

use proc_macro::TokenStream;
use quote::{format_ident, quote};
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input, Ident, LitStr,
};

struct StaticMacro {
    filename: String,
}

impl Parse for StaticMacro {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let filename: LitStr = input.parse()?;

        Ok(StaticMacro {
            filename: filename.value(),
        })
    }
}

#[proc_macro]
pub fn static_array_from_file(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as StaticMacro);
    let filename = input.filename;

    let mut c = env::current_dir().unwrap();
    c.push(&filename);
    let f = fs::OpenOptions::new().read(true).open(&c);
    if f.is_err() {
        let e = format!("cant open file: {}", c.to_string_lossy());
        return quote!(::std::compile_error!(#e);).into();
    }

    let mut buf = BufReader::new(f.unwrap());
    let mut sv = vec![];
    let mut s = String::new();
    while let Ok(sz) = buf.read_line(&mut s) {
        if sz == 0 {
            break;
        }
        sv.push(s.clone().trim().to_owned());
        s.clear();
    }

    let tokens = quote! {
        { &[ #(#sv),* ] }
    };

    tokens.into()
}

struct GenSetMethod {
    name: Ident,
}

impl Parse for GenSetMethod {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let name: Ident = input.parse()?;

        Ok(GenSetMethod { name })
    }
}

#[proc_macro]
pub fn gen_min_max_method(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as GenSetMethod);
    let funcname = format_ident!("set_{}", input.name);
    let name = syn::LitStr::new(&input.name.to_string(), input.name.span());
    let tokens = quote! {
        fn #funcname(mut self, min: Option<i32>, max: Option<i32>) -> Self {
            let v = json!({
                "min": min,
                "max": max,
            });
            let m = self.filters.entry(#name.to_string()).or_default();
            *m = v;
            self
        }
    };
    tokens.into()
}

#[proc_macro]
pub fn gen_option_method(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as GenSetMethod);
    let funcname = format_ident!("set_{}", input.name);
    let name = syn::LitStr::new(&input.name.to_string(), input.name.span());
    let tokens = quote! {
        fn #funcname(mut self, val: bool) -> Self {
            let v = json!({
                "option": val,
            });
            let m = self.filters.entry(#name.to_string()).or_default();
            *m = v;
            self
        }
    };
    tokens.into()
}
