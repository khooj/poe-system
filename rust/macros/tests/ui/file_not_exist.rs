use macros::static_array_from_file;

pub static TYPES: &[&'static str] = static_array_from_file!("file.txt");

fn main() {}