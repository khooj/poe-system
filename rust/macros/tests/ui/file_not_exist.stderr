error: macro expansion ignores token `;` and any following
 --> tests/ui/file_not_exist.rs:3:37
  |
3 | pub static TYPES: &[&'static str] = static_array_from_file!("file.txt");
  |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ caused by the macro expansion here
  |
  = note: the usage of `static_array_from_file!` is likely invalid in expression context

error: cant open file: $WORKSPACE/target/tests/macros/file.txt
 --> tests/ui/file_not_exist.rs:3:37
  |
3 | pub static TYPES: &[&'static str] = static_array_from_file!("file.txt");
  |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `static_array_from_file` (in Nightly builds, run with -Z macro-backtrace for more info)
